### PR TITLE
[flang][Driver] Support -rpath, -shared, and -static in the frontend

### LINKED ISSUE #65546: flang-new: error: unknown argument: '-rpath'
Seen when trying to use flang-new with mpifort:

    flang-new: error: unknown argument: '-rpath'

clang accepts '-rpath'
flang-new should probably accept the same linker options as clang

### PR DESCRIPTION
**Pull request by tarunprabhu**
Enable -rpath, -shared, and -static for the flang frontend. This brings it in line with clang. Fixes issue #65546.

### CODE REVIEW

File: flang/test/Driver/linker-flags.f90 [hunk 1]
```diff
@@ -50,3 +50,19 @@
 ! MSVC-SAME: FortranDecimal.lib
 ! MSVC-SAME: /subsystem:console
 ! MSVC-SAME: "[[object_file]]"
+
+! Verify that certain linker flags are known to the frontend and are passed on
+! to the linker.
+
+! RUN: %flang -### -rpath /path/to/dir %s 2>&1 \
```
Comment (MaskRay) [inline flang/test/Driver/linker-flags.f90:57]: Consider using one command to test many options that can be used together.

Note that you can also use -static -shared together: 408e6de8c09fc7e71329199947e977fe1c40caf1 , so I'm fine if you use one line for -shared and another for -static :)

Comment (MaskRay) [reply flang/test/Driver/linker-flags.f90:57]: The tests are trivial but having too many tests slows down execution. It certainly doesn't matter in this case but if everybody adds more tests than necessary, the testsuite execution will be very slow. In this case, just test all options in one command.

Comment (tarunprabhu) [reply flang/test/Driver/linker-flags.f90]: Fair enough. Done.

File: flang/test/Driver/linker-flags.f90 [hunk 2]
```diff
@@ -50,3 +50,15 @@
...
+! Verify that certain linker flags are known to the frontend and are passed on
+! to the linker.
+
+! RUN: %flang -### -rpath /path/to/dir -shared -static %s 2>&1 \
+! RUN:   | FileCheck --check-prefix=CHECK-LINKER-FLAGS %s
+! CHECK-LINKER-FLAGS: ld
+! CHECK-LINKER-FLAGS-DAG: "-rpath" "/path/to/dir"
+! CHECK-LINKER-FLAGS-DAG: "-shared"
+! CHECK-LINKER-FLAGS-DAG: "-static"
+
```
Comment (MaskRay) [inline flang/test/Driver/linker-flags.f90:63]: Delete trailing blank lines

Flags mean boolean options. Better to rename the prefix.

**Review by MaskRay — APPROVED** [review]

**Review by banach-space — APPROVED** [review]

**Comment by tarunprabhu** [comment]
Thanks @MaskRay and @banach-space for taking a look at this. I have marked the test as unsupported on windows because of a failing buildkite. I don't have access to a Windows machine so I can't add an equivalent test for that platform.

**Comment by banach-space** [comment]
> Thanks @MaskRay and @banach-space for taking a look at this. I have marked the test as unsupported on windows because of a failing buildkite. I don't have access to a Windows machine so I can't add an equivalent test for that platform.

Thanks for working for this! We should investigate the Windows situation rather than disabling this test altogether - a lot of effort has gone into making it work and we should maintain that. Perhaps @mstorsjo or @DavidTruby could help?

**Review by banach-space — CHANGES_REQUESTED** [review]
Please re-enable the tests in linker-flags.f90. 

If these options don't work on Windows then we should try to document it. Also, tests for `-rpath` can be moved to a dedicated file if we need to disable them on a particular platform.

**Comment by MaskRay** [comment]
I think Windows does not work likely because `ld` is absent in PATH. You need `-Bxxx/bin` to specify a directory that contains `ld`, but I don't remember whether `ld.exe` is needed instead.

**Comment by tarunprabhu** [comment]
> I think Windows does not work likely because `ld` is absent in PATH. You need `-Bxxx/bin` to specify a directory that contains `ld`, but I don't remember whether `ld.exe` is needed instead.

I think what I saw on the buildkite was Link.exe, but I don't know if that is always the case.

**Comment by tarunprabhu** [comment]
The buildkite [fails](https://buildkite.com/llvm-project/github-pull-requests/builds/3888#018acca9-e640-47a6-bfe8-3831aeddeb41) yet again because the link flags seem to be mapped differently. 

The link line generated on Windows is:

`"C:\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\Hostx64\\x64\\link.exe" "-out:a.exe" "-libpath:c:\\ws\\src\\build\\lib" "Fortran_main.lib" "FortranRuntime.lib" "FortranDecimal.lib" "/subsystem:console" "-nologo" "-dll" "-implib:a.lib" "-rpath" "/path/to/dir" "C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\lit-tmp-oaf0t3ny\\linker-flags-de9f45.o"`

I don't see `-shared` or `-static`. Can someone help me with this? Is `-dll` is the equivalent of `-shared`? Does windows support static libraries?

**Comment by DavidTruby** [comment]
Sorry I haven’t had a chance to look at this yet but I can help you resolve the Windows issues here. 

Windows does support static libraries and that’s the default for the runtime at the moment, I’ve not actually tried linking the flang runtime dynamically on Windows so that may be the issue. There’s an additional hurdle on Windows as well that libraries statically or dynamically linked against libc are not ABI compatible, so I’m not sure how these flags might interact with that without taking a look

**Comment by tarunprabhu** [comment]
> I'll try to test this patch on a Windows on ARM machine. But it may take some time to setup and build flang on it.

Thanks @luporl. Any help is most appreciated.

**Comment by MaskRay** [comment]
These tests just need a `--target=x86_64-linux-gnu`: which is how clang/test/Driver tests these stuff. Non-ELF platforms such as Apple, Windows, and AIX, may not support some options.

Ideally we should express that the test works for every ELF platform or everything except baremetal, but we don't infrastructure for it. Using a specific target triple is somewhat a convention at least for newer tests.

**Comment by luporl** [comment]
Adding `--target=x86_64-linux-gnu`, as @MaskRay suggests, fixes the test on Windows. But then we are not actually testing linking on Windows.

On my Windows machine, the test output is practically the same as that reported by buildkite, as mentioned by @tarunprabhu above. I see no differences in the output with or without `-static`. Using `-shared` adds `-dll -implib:a.lib` and the output becomes a `dll` instead of an `exe`, which makes sense, according to MSVC documentation (https://learn.microsoft.com/en-us/cpp/build/reference/dll-build-a-dll?view=msvc-170).

I've also tried linking a hello world program with and without `-static` and it runs fine. With `-shared`, it also produces an `a.exe` file (`-o a.dll` is needed to create the output file with the right extension), but that doesn't run, because it's actually a `dll` it seems, although I didn't try to use it as such.

In short, when using MSVC linker, `-rpath` is passed to it, `-static` doesn't change its invocation and `-shared` is translated to `-dll`.

**Comment by kiranchandramohan** [comment]
@banach-space Is this patch OK with you?

File: flang/test/Driver/linker-flags.f90 [hunk 3]
```diff
@@ -48,3 +48,17 @@
...
+! RUN: %flang -### --target=x86_64-linux-gnu -rpath /path/to/dir -shared \
+! RUN:     -static %s 2>&1 | FileCheck \
+! RUN:     --check-prefixes=CHECK-LINKER-OPTIONS,GNU-LINKER-OPTIONS %s
+! RUN: %flang -### --target=x86_64-windows-msvc -rpath /path/to/dir -shared \
+! RUN:     -static %s 2>&1 | FileCheck \
+! RUN:     --check-prefixes=CHECK-LINKER-OPTIONS,MSVC-LINKER-OPTIONS %s
+! CHECK-LINKER-OPTIONS-DAG: "-rpath" "/path/to/dir"
+! GNU-LINKER-OPTIONS-DAG: "-shared"
+! MSVC-LINKER-OPTIONS-DAG: "-dll"
+! GNU-LINKER-OPTIONS-DAG: "-static"
```
Comment (banach-space) [inline flang/test/Driver/linker-flags.f90:64]: I have a few suggestions:
* The tests above this block test for default (i.e. generated by the driver automatically) linker flags when generating an executable. This block does something very different. I would separate the two things into different files. 
* Why would we test `-shared` and `-static` in one invocation?
* We should make sure that what's being tested is indeed the linker invocation. This can be done by e.g. adding something akin `! UNIX-LABEL:  "{{.*}}ld{{(\.exe)?}}"` to make sure that it's indeed the linker invocation that's being matched.

**Comment by banach-space** [comment]
> @banach-space Is this patch OK with you?

Yup, I'm ok with this direction. Made a few small suggestions, but nothing substantial.

**Review by banach-space — APPROVED** [review]
Thanks for seeing this through, that's much appreciated!

File: flang/test/Driver/dynamic-linker.f90
[the remark this answers is not in the document]
Comment (tarunprabhu) [reply flang/test/Driver/dynamic-linker.f90:1]: clang has a [test](https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/dynamic-linker.c) that is similar and in a file with that name.
